### PR TITLE
resolves #13 add iterations annotation to limit iterations

### LINKED ISSUE #13: Iterations annotation
Create an annotation that can cause the spec to run the specified number of iterations. This should most likely be implemented as an AST transformation during the Canonicalization phase (after Semantic Analysis). Spock does it's work during Semantic Analysis so the methods that return the data providers in the where block are present in the AST at that point. We can hook in to the data provider methods and augment the producers to run the desired number of Iterations. I looked at using extensions but there is no good way to hook into the data providers without weird hacks like using a thread local as a flag or asking for (and implementing) a hook for data providers in the spock runtime.

### PR DESCRIPTION
**Pull request by Bijnagte**
@mariogarcia this could use some more tests and documentation. What do you think about the approach though?

### CODE REVIEW

**Comment by mariogarcia** [comment]
Hi @Bijnagte 

Sorry I've been busy this week. Look nice, is definitely better than I had in mind, I didn't realize at `CANONICALIZATION` data providers are extracted as independent methods, good catch! 

The only thing I would argue for is, why a global transformation ? I think it would be more efficient to use a local transformation that is only applied to those classes/methods annotated by `@Iterations` instead of looping through all classes of every module asking which one is a specification.

Anyway I think you're definitely on the good track. I'll try to review it a little bit further tonight to see if I can add anything to it.

Good work Dylan :)

**Comment by Bijnagte** [comment]
The reason I went with a global is because the data provider methods are created by the global spock transformation so they can't be annotated. I could not find a way to access the provider methods from the  feature method since you only get the AST below what is annotated.

There could be a annotation required on the class to enable the transform and still just use the iterations annotation on the features but that would require every Specification to have multiple annotations.

File: src/main/groovy/spock/genesis/transform/GenASTTransformation.groovy [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+    @Override
+    void visit(ASTNode[] nodes, SourceUnit sourceUnit) {
+        ModuleNode module = (ModuleNode) nodes[0]
+        List<ClassNode> classes = module.classes
+
+        for (ClassNode clazz in classes) {
+            if (isSpec(clazz)) {
+                def classIterationsAnnotation = findIterationsAnnotation(clazz)
+                Integer classIterations
```
Comment (Bijnagte) [inline src/main/groovy/spock/genesis/transform/GenASTTransformation.groovy:44]: I don't have a test that shows it yet but Iterations can be on the class also and then overridden for each feature as needed.

**Comment by Bijnagte** [comment]
@mariogarcia I think this is ready to go unless you have anything that you would like to add.

**Comment by mariogarcia** [comment]
I only have a comment on the default number of iterations, if we were using something like quickCheck in Haskell or JUnit quickCheck is always expected 100 iterations. 
- http://pholser.github.io/junit-quickcheck/site/0.6/usage/sample-size.html
- http://www.cs.tufts.edu/~nr/cs257/archive/john-hughes/quick.pdf

My initial feeling would be to stick to that number, as it seems to be the standard when using generative testing. IMO 0 is a bad number. In any case it would be nice to add a mention to it in the documentation. It could be added as a `NOTE:` or `IMPORTANT:` si everybody will know what it means not setting any number.

**Comment by Bijnagte** [comment]
Good call I switched it to 100 and added that to the documentation.

**Pull request merged**